stan/math/rev/fun/divide.hpp

```diff
@@ -10,171 +10,41 @@
-    adjM.adj() += invc_ * adjResult.adj();
-  }
-};
-
-}  // namespace internal
-
-/**
- * Return matrix divided by scalar.
- *
- * @tparam Mat type of the matrix or expression
- * @param[in] m specified matrix or expression
- * @param[in] c specified scalar
- * @return matrix divided by the scalar
- */
-template <typename Mat, typename = require_eigen_vt<std::is_arithmetic, Mat>>
-inline auto divide(const Mat& m, const var& c) {
-  auto* baseVari
-      = new internal::matrix_scalar_divide_dv_vari<Mat::RowsAtCompileTime,
-                                                   Mat::ColsAtCompileTime>(m,
-                                                                           c);
-  Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime> result(
-      m.rows(), m.cols());
-  result.vi()
-      = Eigen::Map<matrix_vi>(baseVari->adjResultRef_, m.rows(), m.cols());
-  return result;
-}
-
-/**
- * Return matrix divided by scalar.
- *
- * @tparam Mat type of the matrix or expression
- * @param[in] m specified matrix or expression
- * @param[in] c specified scalar
- * @return matrix divided by the scalar
- */
-template <typename Mat, typename = require_eigen_vt<is_var, Mat>>
-inline auto divide(const Mat& m, const double& c) {
-  auto* baseVari
-      = new internal::matrix_scalar_divide_vd_vari<Mat::RowsAtCompileTime,
-                                                   Mat::ColsAtCompileTime>(m,
-                                                                           c);
-  Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime> result(
-      m.rows(), m.cols());
-  result.vi()
-      = Eigen::Map<matrix_vi>(baseVari->adjResultRef_, m.rows(), m.cols());
-  return result;
-}
 
 /**
  * Return matrix divided by scalar.
  *
  * @tparam Mat type of the matrix or expression
+ * @tparam Scal type of the scalar
  * @param[in] m specified matrix or expression
  * @param[in] c specified scalar
  * @return matrix divided by the scalar
  */
-template <typename Mat, typename = require_eigen_vt<is_var, Mat>,
-          typename = void>
-inline auto divide(const Mat& m, const var& c) {
-  auto* baseVari
-      = new internal::matrix_scalar_divide_vv_vari<Mat::RowsAtCompileTime,
-                                                   Mat::ColsAtCompileTime>(m,
-                                                                           c);
-  Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime> result(
-      m.rows(), m.cols());
-  result.vi()
-      = Eigen::Map<matrix_vi>(baseVari->adjResultRef_, m.rows(), m.cols());
-  return result;
+template <typename Mat, typename Scal, require_eigen_t<Mat>* = nullptr,
+          require_stan_scalar_t<Scal>* = nullptr,
+          require_any_st_var<Mat, Scal>* = nullptr>
+inline Eigen::Matrix<var, Mat::RowsAtCompileTime, Mat::ColsAtCompileTime>
+divide(const Mat& m, Scal c) {
+  auto arena_m = to_arena_if<!is_constant<Mat>::value>(m);
```

Comment:
I just figured this could cause a memory leak. If `m` contains `double`s, `to_arena_if` returns  a reference to `m`. `arena_m` is captured by the lambda making a copy (it is used in lambda, so it is captured even if that if is not executed. It might be optimized by compiler, but I would not rely on compiler to eliminate potential memory leak). That last copy's destructor will never be called.

I guess that was poor design on my part, as I did the same thing in `dot_product`.

Reply:
I think we need another variant of `to_arena_if` that will return some trivially destructible type (for example empty Map) when the condition is false. Not sure how to name them not to cause confusion, as we also need this version in some cases.

Reply:
Oh yeah, so things copied into the lambda would never have their destructors called? Like the destructor on the vari that holds the lambda is never called -- what would happen if I allocated a std::vector inside a lambda passed to a callback function? Is this what you're getting at?

Reply:
>Oh yeah, so things copied into the lambda would never have their destructors called? Like the destructor on the vari that holds the lambda is never called

Yeah.

> what would happen if I allocated a std::vector inside a lambda passed to a callback function?

Allocating stuff in the lambda is fine. Just capturing things that allocate stuff is a problem.

Reply:
I made an issue for this: https://github.com/stan-dev/math/issues/2025 .

Reply:
Wonder if it's worth me spending a day figuring out how to use the clang address sanitizer stuff to see if we can catch these things in tests